### ISSUE TITLE
valset: fair rule to determine active valset when equal EP/Stake

### ISSUE THREAD

**Issue by alpe** [confio/poe-contracts#22]
With equal EP/ Stake score the (lexicographically) address is taken into account. This can be gamed to be in the active set and may cause frustration and support issues.
Not sure how likely this is on production but on testnets this happened. Let's discuss if there is a simple solution to that so that the active set remains stable.
Other meta data:
* Validator creation time/ block/ tx position
... ?

**Comment by maurolacy** [comment]
A (hashed) combination of block time + block height, perhaps with some extra metadata that can be easily obtained from `Env` and / or `MessageInfo` would make it very difficult to predict, and would be relatively simple to implement.

**Comment by maurolacy** [comment]
Take into account also that this is a side issue, and mostly caused by the testnet setup / preset values. The normal situation will be that not many validators will have exactly the same weight.

**Comment by ethanfrey** [comment]
This is an artifact.

Simply sending random amounts of tokens to each validator to self-stake should be sufficient for the testnet

**Comment by maurolacy** [comment]
Not sure if we want to "fix" this, but was thinking that another option could be adding the block timestamp as  the hidden field. With a negative sign, or after subtracting it from `u64:::MAX`. That way, earlier validators will have precedence when breaking ties.

**Comment by ethanfrey** [comment]
you mean "start time"?

I think we store and return that for the grpc api. we could index with those as well I guess.

but that would be a breaking migration for a future testnet

**Comment by maurolacy** [comment]
Yes. It would need to be negated or subtracted, because it's a descending order, but yes, start_time will do.

**Comment by maurolacy** [comment]
I'm now realising that the validator set results are indexed by (index + primary) key, not value. So, it's not so simple to fix this. But it surely can be done, extending the index definition or so.

**Comment by ethanfrey** [comment]
This is a nice idea, may break our APIs. Does tg4-mixer have access to "start_time" to include in the index?

Happy for a spike to look at what code to change... no tests, doesn't even need to pass CI, just a proof of concept of where it can be modified.

**Comment by maurolacy** [comment]
> This is a nice idea, may break our APIs. Does tg4-mixer have access to "start_time" to include in the index?

Good point, we need to store the height along with the weight when adding (but not updating) a member.

> 
> Happy for a spike to look at what code to change... no tests, doesn't even need to pass CI, just a proof of concept of where it can be modified.

Will do.

**Comment by ethanfrey** [comment]
I had a quick look at #81 and added some comments on the design/architecture.
Thanks for looking into this

**Comment by maurolacy** [comment]
After playing a bit with this, I think the "less bad" (less disruptive) way of doing this (after not doing it at all) would be:

- Add an optional `start_height: Option<u64>` field to both `Member` and `MemberResponse`. Modify contract code to work with this optional field. Notice that Go / tgrade code wouldn't need to me modified, as this change is backwards compatible.
- Modify the `list_member_by_points` impl in `tg4-mixer` (and `tg4-mixer` alone) to work with this field. That means, changing the `members()` `IndexedMap` to store a tuple of `(points, start_height)` values, and returning this `start_height` in the response.

The advantages of this approach are:

- Minimum changes to the code base. Only `tg4-mixer` is modified. This means copying / adapting the `tg4` helpers for `tg4-mixer`, but still.
- Minimum changes to the `tg4` spec. The `tg4` spec is modified, but the change is backwards compatible at the API level.

After this, there will still be a problem with raw queries: The data behind the `MEMBERS` namespace will not be standard anymore, as `tg4-mixer` will store different things under `MEMBERS` than the other group contracts. This will require adapting the `is_member()` raw query to either:

- Work with both data formats. (ugly)
- Provide a different helper implementation for `tg4-mixer` contracts. (uglier)
- Use another way to determine tg4 membership. (workable)
- Store a tuple as part of the MEMBERS data always. Perhaps something like `(u64, Option<u64>)`. (acceptably ugly)

**Comment by ethanfrey** [comment]
I agree with your design above. It sounds solid and well designed, and minority disruptive.

I have just one question about this:

```
Add an optional start_height: Option<u64> field to both Member and MemberResponse. Modify contract code to work with this optional field. Notice that Go / tgrade code wouldn't need to me modified, as this change would be backwards compatible with it.
```

Where do we set this? Who sets this?
Is this just computed in tg4-mixer from Env when adding a member from the hooks?

**Comment by maurolacy** [comment]
> I agree with your design above. It sounds solid and well designed, and minority disruptive.
> 
> I have just one question about this:
> 
> ```
> Add an optional start_height: Option<u64> field to both Member and MemberResponse. Modify contract code to work with this optional field. Notice that Go / tgrade code wouldn't need to me modified, as this change would be backwards compatible with it.
> ```
> 
> Where do we set this? Who sets this? Is this just computed in tg4-mixer from Env when adding a member from the hooks?

Exactly. The other contracts can set it to zero (or `i64::MAX`)

They don't have an associated index for it, etc.
> 
> And this:
> 
> ```
> Store a tuple as part of the MEMBERS data always. Perhaps something like (u64, Option<u64>)?. (acceptably ugly. This will also have the advantage of avoiding the copy of the tg4 state helpers to `tg4-mixer).
> ```
> 
> I kind of grimace using tuples in the json. But this is only state and is only meant to be read by rust code, so we don't have to worry how Go/JS will parse it. So, I guess this is fine as well.

I can create a small `MemberInfo` struct for it. It would have to be part of the `tg4`, though, because even when it's internal, it must be available, as it'll be returned from raw queries.

**Comment by ethanfrey** [comment]
tuple is more efficient I guess?

good point about the standard tg4 storage changes. this just means all contracts must have this change or not have this change (same version of poe-contracts), so best to change it now before mainnet

**Comment by maurolacy** [comment]
Already went with the ancillary `MemberInfo` struct. It has some advantages, like a default and constructors. It's more verbose, but therefore clearer. So, good for documentation, maintenance, extensibility, etc.

**Comment by maurolacy** [comment]
Closed by #118.
